File: src/LanAudioRelay.Core/Protocol/Pairing.cs
using System.Security.Cryptography;
using System.Text;

namespace LanAudioRelay.Core.Protocol;

public static class Pairing
{
    public static string GenerateCode()
    {
        return RandomNumberGenerator.GetInt32(0, 1_000_000).ToString("D6");
    }

    public static string CreateNonce()
    {
        Span<byte> bytes = stackalloc byte[16];
        RandomNumberGenerator.Fill(bytes);
        return Convert.ToBase64String(bytes);
    }

    public static string CreateProof(string code, string clientNonce, string serverNonce)
    {
        var material = $"{NormalizeCode(code)}:{clientNonce}:{serverNonce}";
        var hash = SHA256.HashData(Encoding.UTF8.GetBytes(material));
        return Convert.ToBase64String(hash);
    }

    public static bool VerifyProof(string expectedCode, string clientNonce, string serverNonce, string proof)
    {
        var expected = Encoding.ASCII.GetBytes(CreateProof(expectedCode, clientNonce, serverNonce));
        var actual = Encoding.ASCII.GetBytes(proof);
        return expected.Length == actual.Length && CryptographicOperations.FixedTimeEquals(expected, actual);
    }

    private static string NormalizeCode(string code)
    {
        var digits = new string(code.Where(char.IsDigit).ToArray());
        return digits.PadLeft(6, '0')[^6..];
    }
}

File: mac/LanAudioRelayMac/Sources/LanAudioRelayMacCore/Pairing.swift
import CryptoKit
import Foundation
import Security

public enum Pairing {
    public static func generateCode() -> String {
        String(format: "%06d", Int.random(in: 0..<1_000_000))
    }

    public static func createNonce() -> String {
        var bytes = [UInt8](repeating: 0, count: 16)
        _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        return Data(bytes).base64EncodedString()
    }

    public static func createProof(code: String, clientNonce: String, serverNonce: String) -> String {
        let material = "\(normalizeCode(code)):\(clientNonce):\(serverNonce)"
        let digest = SHA256.hash(data: Data(material.utf8))
        return Data(digest).base64EncodedString()
    }

    public static func verifyProof(
        expectedCode: String,
        clientNonce: String,
        serverNonce: String,
        proof: String
    ) -> Bool {
        createProof(code: expectedCode, clientNonce: clientNonce, serverNonce: serverNonce) == proof
    }

    private static func normalizeCode(_ code: String) -> String {
        let digits = code.filter(\.isNumber)
        if digits.count >= 6 {
            return String(digits.suffix(6))
        }
        return String(repeating: "0", count: 6 - digits.count) + digits
    }
}
